SYCL/ESIMD/api/functional/ctors/common.hpp

```diff
@@ -0,0 +1,165 @@
+
+#include <type_traits>
+
+namespace esimd_test {
+namespace api {
+namespace functional {
+
+// Dummy kernel for submitting some code into device side.
+template <typename DataT, int NumElems, typename T> struct Kernel;
+
+template <typename DataT>
+using allocator_type = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
+
+template <typename DataT>
+using vector_with_allocator = std::vector<DataT, allocator_type<DataT>>;
+
+// Struct that provise methods for constructing inf, lowest, nan and max values
+// for reference data.
+template <typename DataT> struct value {
+  static DataT inf() {
+    DataT inf_val{};
+    if constexpr (std::is_same_v<DataT, sycl::half>) {
+      inf_val = 0;
+    } else {
+      inf_val = std::numeric_limits<DataT>::infinity();
+    }
+    return inf_val;
+  }
+
+  static DataT lowest() {
+    DataT min_val{};
+    if constexpr (std::is_same_v<DataT, sycl::half>) {
+      min_val = 0;
+    } else {
+      min_val = std::numeric_limits<DataT>::lowest();
+    }
+    return min_val;
+  }
+
+  static DataT nan() {
+    DataT nan_val{};
+    if constexpr (std::is_same_v<DataT, double>) {
+      nan_val = sycl::nan(42UL);
+    } else if constexpr (std::is_same_v<DataT, float>) {
+      nan_val = sycl::nan(42U);
+    } else if constexpr (std::is_same_v<DataT, sycl::half>) {
+      nan_val = 0;
+    }
+    return nan_val;
+  }
+
+  static DataT max() {
+    DataT max_val{};
+    if constexpr (std::is_same_v<DataT, sycl::half>) {
+      max_val = 0;
+    } else {
+      max_val = std::numeric_limits<DataT>::max();
+    }
+    return max_val;
+  }
+};
+
+// Generate std::vector with reference according to current tested data type and
+// number of elements.
+template <typename DataT, int NumElems>
+void generate_ref_data(vector_with_allocator<DataT> &ref_data) {
+
+  const DataT min = value<DataT>::lowest();
+  const DataT min_half = min / 2;
+  const DataT max = value<DataT>::max();
+  const DataT max_half = max / 2;
+  const DataT inf = value<DataT>::inf();
+  const DataT min_plus_one = min + 1;
+  const DataT max_minus_one = max - 1;
+
+  if constexpr (std::is_signed_v<DataT>) {
+    if constexpr (NumElems == 1) {
+
+      ref_data.insert(ref_data.end(), {min, min_half, max, max_half, 0});
+    } else {
+      ref_data.insert(ref_data.end(), {min, min_plus_one, min_half, max,
+                                       max_minus_one, max_half, -1, 0});
+      for (size_t it = 8; it < NumElems; it++) {
+        ref_data.push_back(it);
+      }
+    }
+  }
+
+  else if constexpr (std::is_unsigned_v<DataT>) {
+    if constexpr (NumElems == 1) {
+      ref_data.insert(ref_data.end(), {max, max_half, 0});
+    } else {
+      ref_data.insert(ref_data.end(), {max, max_minus_one, max_half, 0});
+      for (size_t it = 4; it < NumElems; it++) {
+        ref_data.push_back(it);
+      }
+    }
+  }
+
+  else if constexpr (std::is_same_v<DataT, sycl::half> ||
+                     std::is_floating_point_v<DataT>) {
```

Comment:
```suggestion
  else if constexpr (type_traits::is_floating_point_v<DataT>) {
```
We might need it sooner or later.. May be it's out of scope for this PR though

Reply:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/578/commits/3f3cc72062810675bd658f89ec1e2ce1478da3a1#diff-f085fe3f94e6fdb4bc2da0518bd492beb702cd3f8181a461be78bcc561ccc514R150)